app/src/cose.h

```diff
@@ -131,25 +300,67 @@ namespace scitt::cose
       throw COSEDecodeError("Failed to decode protected header");
     }
 
-    if (header_items[ALG_INDEX].uDataType == QCBOR_TYPE_NONE)
+    if (header_items[PROFILE_INDEX].uDataType != QCBOR_TYPE_NONE)
     {
-      throw COSEDecodeError("Missing algorithm in protected header");
+      parsed.kid = cbor::as_string(header_items[PROFILE_INDEX].val.string);
     }
-    if (header_items[KID_INDEX].uDataType != QCBOR_TYPE_NONE)
+    if (header_items[ALG_INDEX].uDataType != QCBOR_TYPE_NONE)
     {
-      parsed.kid = cbor::as_string(header_items[KID_INDEX].val.string);
+      parsed.alg = header_items[ALG_INDEX].val.int64;
     }
     if (header_items[ISSUER_INDEX].uDataType != QCBOR_TYPE_NONE)
     {
       parsed.issuer = cbor::as_string(header_items[ISSUER_INDEX].val.string);
     }
+    if (header_items[CRIT_INDEX].uDataType != QCBOR_TYPE_NONE)
+    {
+      parsed.crit = std::vector<std::variant<int64_t, std::string>>();
+      QCBORItem critItem = header_items[CRIT_INDEX];
+      if (critItem.uDataType != QCBOR_TYPE_ARRAY)
+      {
+        throw COSEDecodeError(
+          "Value type of crit in protected header is not of type array");
+      }
+      int lenOfArray = critItem.val.uCount;
+      if (lenOfArray == 0)
+      {
+        throw COSEDecodeError(
+          "Cannot have crit array of length 0 in cose protected header.");
+      }
+      QCBORDecode_EnterArrayFromMapN(&ctx, COSE_HEADER_PARAM_CRIT);
+      for (int i = 0; i < lenOfArray; i++)
+      {
+        QCBORDecode_GetNext(&ctx, &critItem);
+        if (critItem.uDataType == QCBOR_TYPE_TEXT_STRING)
+        {
+          parsed.crit->push_back(std::variant<int64_t, std::string>(
+            std::string(cbor::as_string(critItem.val.string))));
```

Comment:
```suggestion
          parsed.crit->push_back(std::string(cbor::as_string(critItem.val.string))));
```

I don't think this needs to be so explicit.